src/corehost/cli/fxr/fx_resolver.cpp

```diff
@@ -0,0 +1,466 @@
+//     Frameworks are added to the list as they are resolved.
+//     The order in the list is maintained such that the app is always the first and then the framework in their dependency order.
+//     That means that the root framework (typically Microsoft.NETCore.App) is the last.
+//     Frameworks are never removed as there's no operation which would "remove" a framework reference.
+//     Frameworks are never updated in the list. If such operation is required, instead the function returns FrameworkCompatRetry
+//     and the caller will restart the framework resolution process (with new fx_definitions).
+// Return value
+//     Success - all frameworks were successfully resolved and the final (disk resolved) frameworks are in the fx_definitions.
+//     FrameworkCompatRetry - the resolution algorithm needs to restart as some of already processed references has changed.
+//     FrameworkCompatFailure - the resolution failed with unrecoverable error which is due to framework resolution algorithm itself.
+//     FrameworkMissingFailure - the resolution failed because the requested framework doesn't exist on disk.
+//     InvalidConfigFile - reading of a runtime config for some of the processed frameworks has failed.
+StatusCode fx_resolver_t::read_framework(
+    const host_startup_info_t & host_info,
+    const fx_reference_t & override_settings,
+    const runtime_config_t & config,
+    fx_definition_vector_t & fx_definitions)
+{
+    // Loop through each reference and update the list of newest references before we resolve_fx.
+    // This reconciles duplicate references to minimize the number of resolve retries.
+    for (const fx_reference_t& fx_ref : config.get_frameworks())
+    {
+        const pal::string_t& fx_name = fx_ref.get_fx_name();
+        auto temp_ref = m_newest_references.find(fx_name);
+        if (temp_ref == m_newest_references.end())
+        {
+            m_newest_references.insert({ fx_name, fx_ref });
+            m_oldest_references.insert({ fx_name, fx_ref });
+        }
+        else
+        {
+            if (fx_ref.get_fx_version_number() < m_oldest_references[fx_name].get_fx_version_number())
+            {
+                m_oldest_references[fx_name] = fx_ref;
+            }
+        }
+    }
+
+    StatusCode rc = StatusCode::Success;
+
+    // Loop through each reference and resolve the framework
+    for (const fx_reference_t& fx_ref : config.get_frameworks())
+    {
+        const pal::string_t& fx_name = fx_ref.get_fx_name();
+
+        auto existing_framework = std::find_if(
+            fx_definitions.begin(),
+            fx_definitions.end(),
+            [&](const std::unique_ptr<fx_definition_t> & fx) { return fx_name == fx->get_name(); });
+
+        if (existing_framework == fx_definitions.end())
+        {
+            // Perform a "soft" roll-forward meaning we don't read any physical framework folders yet
+            rc = soft_roll_forward(fx_ref, false);
```

Comment:
Personal preference, but I like having the param name as an inline comment when it's not immediately obvious what some value (`false`) being passed is for.

Reply:
Or, instead maybe pull the call out of the if-else and set/pass a variable accordingly and then it should be clear from the variable name.

Reply:
Moving the call out of the if is not nice... added the inline comments. And also added comments on why we're setting it and what it means - as this is probably the trickiest part of the algorithm.